docs/design/proposals/cephfs-fscrypt.md

```diff
@@ -0,0 +1,276 @@
+   accessible in plain text
+ - *protector* (fscrypt) - A single method or secret plus data used to
+   derive a *protector key*. Example: user login passphrase
+ - *protector key* (fscrypt) - A symmetric key derived from an external
+   source. Used by a *policy* to unwrap a *policy key*
+ - *policy key* (fscrypt) - An encryption key passed to the kernel to
+   unlock a directory
+ - *policy* (fscrypt) - A collection of directories protected and unlocked as a unit
+ - *KMS* - Key management system
+
+## User Visible Change
+
+Similar to the existing RADOS Block Device (RBD) encryption support,
+we propose adding encryption support in the configuration and Key Management Service (KMS)
+integration.
+
+In this example, a user may enable encryption using storage class keys similar
+to RBD. Ceph CSI then configures and unlocks the persistent volumes and CephFS
+subvolumes.
+
+Due to the way `fscrypt` stores metadata, subvolumes have a regular root
+directory containing  a `/.fscrypt` directory and a
+`/ceph-csi-encrypted` directory. The first contains `fscrypt` metadata; the
+latter is the fscrypt-enabled directory made that is accessible to pods.
+
+Example configuration using a secrets-based KMS:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: cephfs-storage-encryption-secret
+stringData:
+  encryptionPassphrase: verysecretpassword
+---
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: csi-cephfs-sc-encrypted
+provisioner: cephfs.csi.ceph.com
+parameters:
+  clusterID: <cluster-id>
+  fsName: cephfs
+
+  encrypted: "true"
+  secretName: "cephfs-storage-encryption-secret"
+  encryptionKMSID: "user-ns-secrets-metadata"
+  secretNamespace: "default"
+  encryptionKMSType: "metadata"
+
+  csi.storage.k8s.io/provisioner-secret-name: csi-cephfs-secret
+  csi.storage.k8s.io/provisioner-secret-namespace: default
+  csi.storage.k8s.io/controller-expand-secret-name: csi-cephfs-secret
+  csi.storage.k8s.io/controller-expand-secret-namespace: default
+  csi.storage.k8s.io/node-stage-secret-name: csi-cephfs-secret
+  csi.storage.k8s.io/node-stage-secret-namespace: default
+reclaimPolicy: Delete
+allowVolumeExpansion: true
+mountOptions:
+  - debug
+```
+
+The change will leverage the existing Ceph CSI KMS and support any
+integration now available to RBD encryption
+
+## Implementation
+
+We suggest to leverage the encryption features in Ceph CSI and integrate
+that with `fscrypt`, a Go tool, and the library for key management and configuration
+of the fscrypt kernel feature.
+
+Ceph CSI and `fscrypt` have a lot of overlap between their key management
+features. The Key Management section will go into detail on how and where keys
+are managed.
+
+  - Ceph CSI provides the user facing configuration and access to key
+    management systems
+  - `fscrypt` handles key derivation, storage of wrapped keys and metadata
+
+The current CephFS subvolume root will remain untouched with the exception that
```

Comment:
Subvolume root refers to the root directory of the subvolume. What gets mapped into a pod once mounted.

On my testsetup this would be:
```
$ bin/ceph fs subvolume info a csi-vol-4fa8245c-9b00-11ec-bdd8-eed58c1c7c89 csi
{
...
    "path": "/volumes/csi/csi-vol-4fa8245c-9b00-11ec-bdd8-eed58c1c7c89/32a6c9c4-d83a-482b-8abf-e6b0ea676f3a",
...
}

$ sudo mount -t ceph -o name=admin,secret=.. 192.168.122.1:40687:/volumes/csi/csi-vol-4fa8245c-9b00-11ec-bdd8-eed58c1c7c89/32a6c9c4-d83a-482b-8abf-e6b0ea676f3a /subvolume_root 
```